Source/API/Extras/CBLIncrementalStore.m

```diff
@@ -2144,102 +2150,65 @@ - (void) mergeContextsWithUpdatedIDs: (NSArray*)updatedIDs deletedIDs: (NSArray*
         return;
     }
 
-    NSMutableSet *needRefreshObjects = [NSMutableSet set];
-    
-    NSMutableDictionary* userInfo = [NSMutableDictionary dictionaryWithCapacity: 3];
-    NSMutableSet* updatedEntities = [NSMutableSet set];
-    if (updatedIDs.count > 0) {
-        NSMutableArray* updated = [NSMutableArray arrayWithCapacity: updatedIDs.count];
-        NSMutableArray* inserted = [NSMutableArray arrayWithCapacity: updatedIDs.count];
-
-        for (NSManagedObjectID* moID in updatedIDs) {
-            NSManagedObject* mObj = [strongRootContext objectRegisteredForID: moID];
-            if (mObj) {
-                [updated addObject: mObj];
-            } else {
-                mObj = [strongRootContext objectWithID: moID];
-                [inserted addObject: mObj];
+    NSArray *sortedContextRootBeforeMain = [self.observingManagedObjectContexts.allObjects sortedArrayUsingDescriptors:@[[NSSortDescriptor sortDescriptorWithKey:NSStringFromSelector(@selector(parentContext)) ascending:YES]]];
+
+    for (NSManagedObjectContext *context in sortedContextRootBeforeMain) {
+        [context performBlock:^{
```

Comment:
This doesn't guarantee that the code in root context performBlock: will get executed before the main's performBlock: because both of the block will get executed in their own dispatch queue.

Explicitly call performBlock: on the root context and inside the block, at the end looping to the other context and do the performBlock.